admiral/pkg/clusters/virtualservice_handler.go

```diff
@@ -612,3 +628,196 @@ func deleteVirtualService(ctx context.Context, exist *v1alpha3.VirtualService, n
+// sourceIngressVirtualService is a map of globalFQDN (.mesh) endpoint to VirtualService
+func generateIngressVirtualServiceForDeployment(
+	deployment *k8sAppsV1.Deployment,
+	sourceIngressVirtualService map[string]*v1alpha3.VirtualService) error {
+	if deployment == nil {
+		return fmt.Errorf("deployment is nil")
+	}
+	workloadIdentityKey := common.GetWorkloadIdentifier()
+	cname := common.GetCname(deployment, workloadIdentityKey, common.GetHostnameSuffix())
+	if cname == "" {
+		return fmt.Errorf("cname is empty")
+	}
+	sniHost, err := generateSNIHost(cname)
+	if err != nil {
+		return err
+	}
+	baseVS, err := getBaseVirtualServiceForIngress(cname, sniHost)
+	if err != nil {
+		return err
+	}
+	sourceIngressVirtualService[cname] = baseVS
+	return nil
+}
+
+// generateIngressVirtualServiceForRollout generates the base virtual service for a given rollout
+// and adds it to the sourceIngressVirtualService map
+// sourceIngressVirtualService is a map of globalFQDN (.mesh) endpoint to VirtualService
+// TODO: Generate VS for canary and blue green strategies
+func generateIngressVirtualServiceForRollout(
+	rollout *argo.Rollout,
+	sourceIngressVirtualService map[string]*v1alpha3.VirtualService) error {
+	if rollout == nil {
+		return fmt.Errorf("rollout is nil")
+	}
+	workloadIdentityKey := common.GetWorkloadIdentifier()
+	cname := common.GetCnameForRollout(rollout, workloadIdentityKey, common.GetHostnameSuffix())
+	if cname == "" {
+		return fmt.Errorf("cname is empty")
+	}
+	sniHost, err := generateSNIHost(cname)
+	if err != nil {
+		return err
+	}
+	baseVS, err := getBaseVirtualServiceForIngress(cname, sniHost)
+	if err != nil {
+		return err
+	}
+	sourceIngressVirtualService[cname] = baseVS
+	return nil
+}
+
+// generateSNIHost generates the SNI host for the virtual service in the format outbound_.80_._.<fqdn>
+// Example: outbound_.80_._.httpbin.global.mesh
+func generateSNIHost(fqdn string) (string, error) {
+	if fqdn == "" {
+		return "", fmt.Errorf("fqdn is empty")
+	}
+	return fmt.Sprintf("outbound_.%d_._.%s", common.DefaultServiceEntryPort, fqdn), nil
+}
+
+// addUpdateVirtualServicesForSourceIngress adds or updates the cross-cluster routing VirtualServices exported to
+// istio-system namespace.
+func addUpdateVirtualServicesForSourceIngress(
+	ctx context.Context,
+	ctxLogger *log.Entry,
+	remoteRegistry *RemoteRegistry,
+	sourceServices map[string]map[string]*k8sV1.Service,
+	sourceIngressVirtualService map[string]*v1alpha3.VirtualService,
+	sourceDeployment map[string]*k8sAppsV1.Deployment,
+	sourceRollout map[string]*argo.Rollout) error {
+
+	if remoteRegistry == nil {
+		return fmt.Errorf("remoteRegistry is nil")
+	}
+
+	for sourceCluster, serviceInstance := range sourceServices {
+
+		destinationHostPort := make(map[string]uint32)
+		if serviceInstance[common.Deployment] != nil {
+			host := serviceInstance[common.Deployment].Name + "." + serviceInstance[common.Deployment].Namespace + ".svc.cluster.local"
```

Comment:
may be i missed in previous review, can we add a dot in the end :) same for rollout